File: app/src/main/java/com/elitec/alejotaller/feature/product/domain/caseUse/ReleaseSoftHoldCaseUse.kt
package com.elitec.alejotaller.feature.product.domain.caseUse

import android.util.Log
import com.elitec.alejotaller.feature.product.domain.repository.ProductRepository
import com.elitec.shared.sale.feature.sale.domain.entity.Sale

/**
 * Rollback soft-hold: reserved -= qty por línea (cancelación / rechazo).
 * Clamp reserved >= 0. Devuelve productIds tocados para señal stock:changed.
 */
class ReleaseSoftHoldCaseUse(
    private val repository: ProductRepository
) {
    suspend operator fun invoke(sale: Sale): Result<List<String>> = runCatching {
        val touched = linkedSetOf<String>()
        for (item in sale.products) {
            val qty = item.quantity.coerceAtLeast(0)
            if (qty == 0) continue
            repository.decrementReserved(item.productId, qty)
                ?: error("No se pudo liberar soft-hold de ${item.productId}")
            touched += item.productId
            Log.i(TAG, "event=soft_hold_released productId=${item.productId} qty=$qty saleId=${sale.id}")
        }
        touched.toList()
    }

    companion object {
        private const val TAG = "ReleaseSoftHold"
    }
}

File: web/src/core/feature/product/domain/caseuse/ReleaseSoftHoldCaseUse.ts
import type { Sale } from "../../../sale/domain/entity/Sale";
import type { ProductRepository } from "../repository/product.repository";

/**
 * Rollback de soft-hold: reserved -= quantity por línea.
 * Usado cuando el pedido se cancela (cliente) o se rechaza (operador / DELETED).
 * Idempotente a nivel de llamada: clamp reserved >= 0.
 */
export class ReleaseSoftHoldCaseUse {
    constructor(private readonly repository: ProductRepository) {}

    async execute(sale: Sale): Promise<string[]> {
        const touched: string[] = [];

        for (const item of sale.products ?? []) {
            const qty = Math.max(0, Math.floor(item.quantity));
            if (qty === 0 || !item.productId) continue;

            const product = await this.repository.getById(item.productId);
            if (!product) {
                console.warn(
                    `[ReleaseSoftHold] producto ausente productId=${item.productId} saleId=${sale.id}`
                );
                continue;
            }

            const nextReserved = Math.max(0, (product.reserved ?? 0) - qty);

            if (import.meta.env.DEV) {
                console.info(
                    `[ReleaseSoftHold] product=${item.productId} reserved ${product.reserved ?? 0} → ${nextReserved}`
                );
            }

            await this.repository.update(item.productId, { reserved: nextReserved });
            touched.push(item.productId);
        }

        return [...new Set(touched)];
    }
}
